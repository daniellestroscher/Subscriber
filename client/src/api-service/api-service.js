const ApiService = {};

const brandToken = process.env.REACT_APP_BRAND_API_TOKEN;
const baseApiURL = process.env.REACT_APP_BRAND_API_BASE_URL;
const baseURL = process.env.REACT_APP_BASE_URL;

ApiService.getAllSubs = async () => {
 let res = await fetch(`${baseURL}/subscriptions`);
 res = await res.json();
 return res;
}


ApiService.postNewSub = async (subscription) => {
  let res = await fetch(`${baseURL}/subscriptions`, {
    method: 'POST',
    body: JSON.stringify(subscription),
    headers: {
      'Content-Type': 'application/json; charset=UTF-8',
    }
  })
  res = res.json();
  return res;
}

ApiService.editSub = async (updatedSub) => {
  let res = await fetch(`${baseURL}/subscriptions`, {
    method: 'PUT',
    body: JSON.stringify(updatedSub),
    headers: {
      'Content-Type': 'application/json; charset=UTF-8',
    }
  })
  res = res.json();
  return res;
}

ApiService.deleteSub = async (uid) => {
    await fetch(`${baseURL}/subscriptions`, {
    method: 'DELETE',
    body: JSON.stringify({id: uid}),
    headers: {
      'Content-Type': 'application/json; charset=UTF-8',
    }
  })
}

ApiService.postSubNotification = async (notificationData) => {
  await fetch(`${baseURL}/notify`, {
    method: 'POST',
    body: JSON.stringify(notificationData),
    headers: {
      'Content-Type': 'application/json; charset=UTF-8',
    }
  })
}

ApiService.putToken = async (token, uid) => {
  await fetch(`${baseURL}/user-token`, {
    method: 'PUT',
    body: JSON.stringify({token: token, userId: uid}),
    headers: {
      'Content-Type': 'application/json; charset=UTF-8',
    }
  })
}

ApiService.getSubOptions = async (domain) => {
  let res = await fetch(`${baseApiURL}${domain}`, {
    method: 'GET',
    headers: {
      'Authorization': 'Bearer' + brandToken,
    }
  });
  res = await res.json();
  return res;
}

export default ApiService;