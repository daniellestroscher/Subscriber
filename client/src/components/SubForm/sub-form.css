.form-cont {
  background-color: rgba(255, 255, 255, .9);
  backdrop-filter: blur(2px);
  box-shadow: 0px 2px 10px 2px #fcbebe;
  border-radius: 5px;
  margin: 5%;
  padding: 20px;
  width: 65vw;
  max-width: 700px;
  align-self: center;
  display: flex;
  justify-content: center;
}

.form-inputs {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 90%;
}

.add-sub-header {
  position: relative;
  display: flex;
  justify-content: flex-start;
  width: 100%;
  align-items: center;
  margin: 15px;
}

.icon-display {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 8rem;
  aspect-ratio: 1 / 1;
  border-radius: 50%;
  cursor: pointer;
  margin-right: 10px;
}

.icon {
  height: 100%;
  width: 100%;
  border-radius: 50%;
  box-shadow: 0 5px 10px -5px rgba(0,0,0,0.5);
}

.back-btn {
  position: absolute;
  color: blueviolet;
  top: 0;
  right: 0;
}

.form-input-box {
  width: 100%;
}
.cycle-select {
  border: none;
  border-radius: 5px;
  margin-left: 10px;
  font-family: 'Dosis', sans-serif;
  font-size: 16px;
  font-weight: 400;
  color: blue;
}

.form-body {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
}

.input-label {
  display: flex;
  font-size: 20px;
  margin-top: 0.5rem;
}

.category-select {
  font-size: 14px;
  border: none;
  border-radius: 5px;
  padding: 10px;
  margin: 10px;
  width: 104%;
  color: gray;
}

.submit-form-btn {
  background-color: transparent;
  border: 1.5px solid rgb(121, 28, 227);
  border-radius: 5px;
  color: rgb(121, 28, 227);
  margin: 2rem;
  padding: 0.7rem;
  font-family: 'Dosis', sans-serif;
  font-weight: 400;
  font-size: 16px;
  cursor: pointer;
  box-shadow: 0 5px 10px -5px rgba(0,0,0,0.5);
}

.submit-form-btn:active {
  box-shadow: none;
}