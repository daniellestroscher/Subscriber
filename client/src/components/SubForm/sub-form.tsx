import React, { useState } from "react";
import "./sub-form.css";
import { Link, useNavigate } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronLeft } from "@fortawesome/free-solid-svg-icons";
import { Notification, Subscription } from "../../Types";
import {
  postSubNotification,
  deleteSub,
  editSub,
  postNewSub,
} from "../../api-service/api-service";
import SubFormItem from "../SubFormItem/sub-form-item";
import { getMessageToken } from "../../firebase";

type Props = {
  subscription?: Subscription;
  setSubs: React.Dispatch<React.SetStateAction<Subscription[] | undefined>>;
  subscriptions: Subscription[];
};

function SubForm({ subscription, setSubs, subscriptions }: Props) {
  let initialState: Subscription;
  let apiServiceMethod: (formData: Subscription) => void;

  if (subscription) {
    initialState = subscription;
    apiServiceMethod = editSub;
  } else {
    initialState = {
      icon: "./add-image-icon.png",
      price: 0,
      title: "",
      start: "",
      cycle: "Monthly",
      category: "",
      reminderDate: "",
      prettyStart: "",
    };
    apiServiceMethod = postNewSub;
  }

  const [imageFile, setImageFile] = useState<File | undefined>();
  const [sub, setSub] = useState(initialState);
  const navigate = useNavigate();
  const imageUploader = React.useRef<HTMLInputElement>(null);

  const uploadImage = async () => {
    const data = new FormData();
    data.append("file", imageFile || "");
    data.append("upload_preset", process.env.REACT_APP_CLOUDINARY_PRESET || "");
    data.append(
      "cloud_name",
      process.env.REACT_APP_CLOUDINARY_CLOUD_NAME || ""
    );
    let imgData = await fetch(process.env.REACT_APP_CLOUDINARY_URL || "", {
      method: "POST",
      body: data,
    });
    imgData = await imgData.json();
    return imgData;
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    let dateTime = new Date(sub.start);
    const prettyStart = dateTime.toDateString();
    let icon = sub.icon;
    if (imageFile) icon = (await uploadImage()).url;
    const subscriptionData = {
      icon,
      price: sub.price,
      title: sub.title,
      category: sub.category,
      start: sub.start,
      prettyStart,
      cycle: sub.cycle,
      reminderDate: new Date(sub.reminderDate).toISOString(),
      _id: subscription?._id,
    };
    apiServiceMethod(subscriptionData);

    if (sub.reminderDate) {
      const delay = new Date(sub.reminderDate).getTime() - new Date().getTime();
      const token = (await getMessageToken()) || "";
      const notification: Notification = {
        token,
        title: sub.title,
        price: sub.price,
        delay: delay,
      };
      postSubNotification(notification);
    }
    setSubs([...subscriptions, subscriptionData]);
    //TODO:change setsubs to refetch saved data
    navigate("/");
  };

  const handleDelete = () => {
    if (subscription?._id) deleteSub(subscription._id.toString());
    let subsMinusDelete = subscriptions.filter(
      (subs) => subs._id !== subscription?._id
    );
    setSubs([...subsMinusDelete]);
    navigate("/");
  };

  return (
    <div className="form-cont">
      <form onSubmit={handleSubmit} className="form-inputs">
        <section className="add-sub-header">
          <input
            type="file"
            accept="image/*"
            multiple={false}
            onChange={(e) => {
              setSub({ ...sub, icon: URL.createObjectURL(e.target.files![0]) });
              setImageFile(e.target.files![0]);
            }}
            ref={imageUploader}
            style={{ display: "none" }}
          ></input>
          <div
            className="icon-display"
            onClick={() => {
              if (imageUploader.current) {
                imageUploader.current.click();
              }
            }}
          >
            <img className="icon" src={sub.icon} alt="Add Icon" />
          </div>
          <div className="payment-info">
            <SubFormItem
              label="Price:"
              placeholder="Price"
              data={sub.price}
              onChange={(e) =>
                setSub({ ...sub, price: parseInt(e.target.value) })
              }
              type="number"
              hasSection={false}
            />
            <select
              className="cycle-select"
              name="cycle"
              onChange={(e) => setSub({ ...sub, cycle: e.target.value })}
            >
              <option value="Monthly">Monthly</option>
              <option value="Annually">Annually</option>
            </select>
          </div>
          <Link to="/">
            <FontAwesomeIcon icon={faChevronLeft} className="back-btn" />
          </Link>
        </section>

        {/* FORM BODY */}
        <div className="form-body">
          <SubFormItem
            label="Title: "
            placeholder="Title"
            data={sub.title}
            onChange={(e) => setSub({ ...sub, title: e.target.value })}
            type="string"
          />
          <select
            className="category-select"
            name="category"
            onChange={(e) => setSub({ ...sub, category: e.target.value })}
          >
            <option value="select">Select Category</option>
            <option value="entertainment">entertainment</option>
            <option value="education">education</option>
            <option value="work">work</option>
            <option value="home">home</option>
            <option value="food">food</option>
            <option value="other">other</option>
          </select>
          <SubFormItem
            label="First Payment: "
            placeholder="First Payment Date"
            data={new Date(sub.start).toLocaleDateString("en-ca")}
            onChange={(e) => setSub({ ...sub, start: e.target.value })}
            type="date"
          />
          <SubFormItem
            label="Remind Me: "
            placeholder={"Reminder Date"}
            data={sub.reminderDate.slice(0, -8)}
            onChange={(e) =>
              setSub({
                ...sub,
                reminderDate: new Date(e.target.value).toISOString(),
              })
            }
            type="datetime-local"
            min={new Date().toISOString().slice(0, -8)}
          />
          <section>
            {subscription ? (
              <button
                className="submit-form-btn"
                type="button"
                onClick={() => handleDelete()}
              >
                Delete Subscription
              </button>
            ) : null}
            <button className="submit-form-btn" type="submit">
              {subscription ? "Edit" : "Add"} Subscription
            </button>
          </section>
        </div>
      </form>
    </div>
  );
}

export default SubForm;
