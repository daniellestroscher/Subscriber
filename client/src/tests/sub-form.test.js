import React from 'react';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import SubForm from '../components/sub-form';
import { mockSubItem } from './mocks';
import { editSub, postNewSub } from '../api-service/api-service'
import renderer from 'react-test-renderer';


describe('SubItem Tests', () => {
  // Date.now.mockImplementation(() => new Date(Date.UTC(2022, 11, 21, 19)).valueOf())
  let subscription;

  beforeEach(() => {
    Date.now = jest.fn(() => Date.UTC(2022, 11, 21, 19))
    subscription = { ...mockSubItem[0] }
  });

  it('Should render the an empty form if no subscription is passed', async () => {
    const tree = renderer
      .create(
        <Router>
          <SubForm apiServiceMethod={postNewSub} />
        </Router>
      )
      .toJSON()
    expect(tree).toMatchSnapshot();
  })

  it('Should render the filled form if no subscription is passed', async () => {
    const tree = renderer
      .create(
        <Router>
          <SubForm apiServiceMethod={editSub} subscription={subscription} />
        </Router>
      )
      .toJSON()
    expect(tree).toMatchSnapshot();
  })
})