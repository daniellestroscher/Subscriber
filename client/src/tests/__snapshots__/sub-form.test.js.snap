// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubItem Tests Should render the an empty form if no subscription is passed 1`] = `
<div
  className="form-cont"
>
  <form
    onSubmit={[Function]}
  >
    <section
      className="add-sub-header"
    >
      <input
        accept="image/*"
        multiple={false}
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <div
        className="icon-display"
        onClick={[Function]}
      >
        <img
          alt="Add Icon"
          className="icon"
          src="./add-image-icon.png"
        />
      </div>
      <label
        className="form-input-label"
      >
        $
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        onChange={[Function]}
        type="number"
        value={0}
      />
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-left back-btn"
          data-icon="chevron-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 278.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </section>
    <div
      className="form-body"
    >
      <label
        className="form-input-label"
      >
        Title: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        onChange={[Function]}
        type="string"
        value="Title here"
      />
      <label
        className="form-input-label"
      >
        First Payment: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        onChange={[Function]}
        type="date"
        value=""
      />
      <label
        className="form-input-label"
      >
        Cycle: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        list="cycleData"
        onChange={[Function]}
        type="text"
        value="Choose Cycle"
      />
      <datalist
        id="cycleData"
      >
        <option
          value="Monthly"
        />
        <option
          value="Annually"
        />
      </datalist>
      <label
        className="form-input-label"
      >
        Remind Me: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        min="2022-12-21T19:00:00.000Z"
        onChange={[Function]}
        type="datetime-local"
        value=""
      />
      <section>
        <button
          className="submit-form-btn"
          onClick={[Function]}
          type="button"
        >
          Delete Subscription
        </button>
        <button
          className="submit-form-btn"
          type="submit"
        >
          Edit Subscription
        </button>
      </section>
    </div>
  </form>
</div>
`;

exports[`SubItem Tests Should render the filled form if no subscription is passed 1`] = `
<div
  className="form-cont"
>
  <form
    onSubmit={[Function]}
  >
    <section
      className="add-sub-header"
    >
      <input
        accept="image/*"
        multiple={false}
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <div
        className="icon-display"
        onClick={[Function]}
      >
        <img
          alt="Add Icon"
          className="icon"
          src="https://cdn.icon-icons.com/icons2/3053/PNG/512/netflix_macos_bigsur_icon_189917.png"
        />
      </div>
      <label
        className="form-input-label"
      >
        $
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        onChange={[Function]}
        type="number"
        value={12}
      />
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-left back-btn"
          data-icon="chevron-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 278.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </section>
    <div
      className="form-body"
    >
      <label
        className="form-input-label"
      >
        Title: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        onChange={[Function]}
        type="string"
        value="Netflix"
      />
      <label
        className="form-input-label"
      >
        First Payment: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        onChange={[Function]}
        type="date"
        value={2022-11-22T01:26:29.807Z}
      />
      <label
        className="form-input-label"
      >
        Cycle: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        list="cycleData"
        onChange={[Function]}
        type="text"
        value="Monthly"
      />
      <datalist
        id="cycleData"
      >
        <option
          value="Monthly"
        />
        <option
          value="Annually"
        />
      </datalist>
      <label
        className="form-input-label"
      >
        Remind Me: 
      </label>
      <input
        className="form-input-box"
        data-testid="form-input-box"
        min="2022-12-21T19:00:00.000Z"
        onChange={[Function]}
        type="datetime-local"
        value={2022-11-22T01:26:29.807Z}
      />
      <section>
        <button
          className="submit-form-btn"
          onClick={[Function]}
          type="button"
        >
          Delete Subscription
        </button>
        <button
          className="submit-form-btn"
          type="submit"
        >
          Edit Subscription
        </button>
      </section>
    </div>
  </form>
</div>
`;
